core/src/Cabana_Parallel.hpp

```diff
@@ -27,24 +27,23 @@ namespace Impl
 {
 
 // No work tag was provided so call without a tag argument.
-template <class WorkTag, class FunctorType, class IndexType>
+template <class WorkTag, class FunctorType, class... IndexTypes>
 KOKKOS_FORCEINLINE_FUNCTION
     typename std::enable_if<std::is_same<WorkTag, void>::value>::type
-    functorTagDispatch( const FunctorType &functor, const IndexType s,
-                        const IndexType a )
+    functorTagDispatch( const FunctorType &functor, IndexTypes &&... indices )
 {
-    functor( s, a );
+    functor( std::forward<IndexTypes>( indices )... );
 }
 
 // The user gave us a tag so call the version using that.
-template <class WorkTag, class FunctorType, class IndexType>
+template <class WorkTag, class FunctorType, class... IndexTypes>
 KOKKOS_FORCEINLINE_FUNCTION
     typename std::enable_if<!std::is_same<WorkTag, void>::value>::type
-    functorTagDispatch( const FunctorType &functor, const IndexType s,
-                        const IndexType a )
+    functorTagDispatch( const FunctorType &functor,
+                        const IndexTypes &&... indices )
```

Comment:
I think you need to drop this `const`, and that's what's causing most your issues (you broke the param pack and type deduction)

Could you also catch this branch up to master please?

Reply:
Thanks for finding that, my bad

Reply:
No worries, glad it fixed it. The compiler really doesn't make it easy to find something like that..